### PR TITLE
Hash based rangelist

### PR DESCRIPTION
**Pull request by subdavis**
@sgratzl this pr includes a few other suggested fixes.  PTAL.  If you disagree, I can revert anything you think is necessary.  Once this is merged, I'll approve #264

### CODE REVIEW

**Comment by sgratzl** [comment]
why don't use use a regular https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set? Morever, with a set based approach you loose the order of selection in contrast to a list

**Comment by subdavis** [comment]
Yep, that would make way more sense.

**Comment by subdavis** [comment]
@sgratzl turns out there is a reason not to use a Set. I found https://github.com/vuejs/vue/issues/2410#issuecomment-318487855

> Sets and Maps are not reactive in vue

I think this implementation is simple enough for now.  We'd have to wrap a set with some other class to get the `first` and `last` properties.

File: web/tests/rangelist.spec.js [hunk 1]
```diff
@@ -1,26 +1,6 @@
 import RangeList from '../src/utils/rangelist';
 
 describe('RangeList', () => {
-  it('Produces non-overlapping ranges', () => {
```
Comment (sgratzl) [inline web/tests/rangelist.spec.js]: why did you delete all those tests?

Comment (subdavis) [reply web/tests/rangelist.spec.js]: Because of the implementation change, this test case no longer applies.  Overlap is not meaningful for a hash-based approach:

Before, `this.members = [ 1, 2, 2, 3, 4, 5 ]` was a possible edge case this test guarded against.

File: web/src/store/index.js [hunk 2]
```diff
@@ -227,7 +227,7 @@ const mutations = {
     } else if (event.ctrlKey && axis === type) {
       ranges.add(idx);
     } else if (!event.ctrlKey && !event.shiftKey) {
-      state.datasets[key].selected.ranges = new RangeList([idx]);
+      state.datasets[key].selected.ranges = new RangeList({ [idx]: true });
```
[the remark this answers is not in the document]
Comment (subdavis) [reply web/src/store/index.js]: Agreed.

File: web/src/utils/rangelist.js [hunk 3]
```diff
@@ -1,29 +1,15 @@
...
-      if (this.members[midIndex] === el) {
-        return midIndex;
-      } if (this.members[midIndex] < el) {
-        lowIndex = midIndex + 1;
-      } else {
-        highIndex = midIndex - 1;
-      }
-    } return null;
+  get members() {
+    return Object.keys(this.members_).map(m => parseInt(m, 10));
```
[the remark this answers is not in the document]
Comment (subdavis) [reply web/src/utils/rangelist.js]: The usages don't need sorting, but I still think that's a reasonable expectation.

File: web/src/utils/rangelist.js [hunk 4]
```diff
@@ -1,29 +1,18 @@
...
-      } if (this.members[midIndex] < el) {
-        lowIndex = midIndex + 1;
-      } else {
-        highIndex = midIndex - 1;
-      }
-    } return null;
+  get members() {
+    return Object.keys(this.members_)
+      .map(m => parseInt(m, 10))
+      .sort();
```
Comment (sgratzl) [inline web/src/utils/rangelist.js]: this will sort alphabetically not numerically

File: web/tests/rangelist.spec.js (hunk 1 above)
Comment (sgratzl) [reply web/tests/rangelist.spec.js]: imo this is still a valid test. I don't get why it shouldn't pass anymore. And the case you are describing is still part of the specification of the RangeList class.
